src/tools/snapshot/nodes/BlockSerializationNode.java

```diff
@@ -0,0 +1,168 @@
+package tools.snapshot.nodes;
+
+import java.nio.ByteBuffer;
+
+import com.oracle.truffle.api.Truffle;
+import com.oracle.truffle.api.frame.FrameDescriptor;
+import com.oracle.truffle.api.frame.FrameSlot;
+import com.oracle.truffle.api.frame.MaterializedFrame;
+
+import som.compiler.Variable.Internal;
+import som.interpreter.FrameOnStackMarker;
+import som.interpreter.Types;
+import som.vm.Symbols;
+import som.vm.constants.Classes;
+import som.vmobjects.SAbstractObject;
+import som.vmobjects.SBlock;
+import som.vmobjects.SInvokable;
+import tools.snapshot.SnapshotBackend;
+import tools.snapshot.SnapshotBuffer;
+
+
+public class BlockSerializationNode extends AbstractSerializationNode {
+  private static final int SINVOKABLE_SIZE = Short.BYTES;
+
+  @Override
+  public void serialize(final Object o, final SnapshotBuffer sb) {
+    assert o instanceof SBlock;
+    SBlock block = (SBlock) o;
+
+    MaterializedFrame mf = block.getContextOrNull();
+
+    if (mf == null) {
+      int base = sb.addObject(block, Classes.blockClass, SINVOKABLE_SIZE + 2);
+      SInvokable meth = block.getMethod();
+      String sid = meth.getIdentifier();
+      sb.putShortAt(base, Symbols.symbolFor(sid).getSymbolId());
+      sb.putShortAt(base + 2, (short) 0);
+    } else {
+      FrameDescriptor fd = mf.getFrameDescriptor();
+
+      Object[] args = mf.getArguments();
+
+      int start = sb.addObject(block, Classes.blockClass,
+          SINVOKABLE_SIZE + ((args.length + fd.getSlots().size()) * Long.BYTES) + 2);
+      int base = start;
+
+      SInvokable meth = block.getMethod();
+      String sid = meth.getIdentifier();
+      sb.putShortAt(base, Symbols.symbolFor(sid).getSymbolId());
+      sb.putByteAt(base + 2, (byte) args.length);
+      base += 3;
+
+      for (int i = 0; i < args.length; i++) {
+        Types.getClassOf(args[i]).getSerializer().serialize(args[i], sb);
+        sb.putLongAt(base + (i * Long.BYTES), sb.getObjectPointer(args[i]));
+      }
+
+      base += (args.length * Long.BYTES);
+
+      int j = 0;
+
+      sb.putByteAt(base, (byte) fd.getSlots().size());
+      base++;
+      for (FrameSlot slot : fd.getSlots()) {
+        // assume this is ordered by index
+        assert slot.getIndex() == j;
+
+        Object value = mf.getValue(slot);
+        switch (fd.getFrameSlotKind(slot)) {
```

Comment:
Hm, perhaps an opportunity for nodes to optimize, too.
Am also wondering whether the normal local variable read nodes can be reused, not sure it's worth it though.

Also note that we do not support all kinds, we don't have byte, float, and int values for instance.